File: src/services/AuthService.ts
import { api } from '@/constants/Api'
import { RequestMethods } from '@/constants/RequestMethods'
import { User } from '@/entities/User'
import { FetchService } from './FetchService'
export namespace AuthService {
  export interface createUserArg {
    username: User['username']
    email: User['email']
    password: User['password']
  }
  export interface loginArg {
    username: User['username']
    password: User['password']
  }
  export interface refreshTokenArg {
    token: User['token']
  }
}
export class AuthService {
  async createUser(arg: AuthService.createUserArg) {
    const url = api.users.create
    return await FetchService.request(arg, RequestMethods.post, url)
  }
  async login(arg: AuthService.loginArg) {
    const url = api.users.login

    return await FetchService.request(arg, RequestMethods.post, url)
  }
  async refreshToken(arg: AuthService.refreshTokenArg) {
    const url = api.users.refreshToken
    return await FetchService.request(arg, RequestMethods.post, url, false)
  }
}

File: packages/remote-kanban-flutter1/lib/services/AuthService.dart
import 'dart:async';

import 'package:dio/dio.dart';
import 'package:test_fl_kanban/constants/Api.dart';
import 'package:test_fl_kanban/constants/RequestMethods.dart';
import 'package:test_fl_kanban/entities/CommonError.dart';
import 'package:test_fl_kanban/entities/User.dart';
import 'package:test_fl_kanban/services/FetchService.dart';

class AuthService {
  Future<UserError> login(User user) async {
    try {
      var response = await FetchService.fetch(
          body: {'username': user.username, 'password': user.password},
          requestMethod: RequestMethods.post,
          url: Api.usersLogin);

      //TODO: save token to shared preferences
      String token = response.data['token'];
      if (token == null) throw CommonError(message: 'Token is null!');
      TokenMap.appLivingToken = AuthService.createToken(token);
      return null;
    } on CommonError catch (e) {
      return UserError(commonError: e);
    }
  }

  Future<Response<dynamic>> refreshToken({String token}) async {
    return await FetchService.fetch(
        body: {'token': token},
        requestMethod: RequestMethods.post,
        url: Api.usersRefreshToken,
        checkToken: false);
  }

  // TODO: add shared preferences
  static Future<String> getLocalToken() async => TokenMap.appLivingToken;

  static String cleanToken(String token) => token.replaceAll('JWT ', '');
  static String createToken(String token) => 'JWT $token';
}
